### PR TITLE
Feature/mapbinding

### LINKED ISSUE #135: Add a data driven / MVVM capability
This is needed to drive proper list / table capabilities.
Also possible data binding like more advanced forms...

This will require a fair bit of discussion and desigm (document to be linked)

### PR DESCRIPTION
**Pull request by andydotxyz**
Complete map and struct binding.
Finishes data binding work for now (ListList and MapList could follow later).

Fixes #135

### Checklist:

- [x] Tests included.
- [x] Lint and formatter run with no errors.
- [x] Tests all pass.
- [x] Public APIs match existing style.

### CODE REVIEW

File: data/binding/mapbinding.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+	for k := range b.val {
+		ret[i] = k
+		i++
+	}
+
+	return ret
+}
+
+func (b *mapBase) Delete(key string) {
+	b.val[key] = nil
```
[the remark this answers is not in the document]
Comment (stuartmscott) [reply data/binding/mapbinding.go]: Please add a test

File: data/binding/gen.go [hunk 2]
```diff
@@ -246,8 +246,8 @@ func Bind{{ .Name }}List(v *[]{{ .Type }}) {{ .Name }}List {
 
 	b := &bound{{ .Name }}List{val: v}
 
-	for _, i := range *v {
-		b.appendItem(Bind{{ .Name }}(&i))
+	for i := range *v {
+		b.appendItem(Bind{{ .Name }}(&((*v)[i])))
```
[the remark this answers is not in the document]
Comment (andydotxyz) [reply data/binding/gen.go:250]: This has been reverted, I have no idea why the github diff is not showing it as removed, sorry,

File: data/binding/mapbinding.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply data/binding/mapbinding.go]: Sorry, I missed the test comment before, quite right. 
Fixed

File: data/binding/gen.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (andydotxyz) [reply data/binding/gen.go]: Ah, @Jacalz found the problem, there were two different fixes floating around. Resolved now

**Review by stuartmscott — APPROVED** [review]

**Pull request merged**